cloud_storage: avoid possible double stop in segment reader

andrwng opened the pull request:
It's possible that when a client disconnects, the abort source subscription associated with the client connection ends up evicting the current segment reader.

1. Remote partition reader is in the process of reading a segment
2. The read eventually scans past the LSO, and stops. This stop() is asynchonous, closing IO streams, etc.
3. The Kafka client disconnects, triggering the abort source subscription callback to call evict_segment_reader(), moving the segment reader to the remote partition's eviction list
4. The remote partition's eviction loop sees this reader and runs stop(), and destructs the underlying parser
5. The read fiber, still stopping, resumes control, but the underlying parser is destructed

This is roughly the sequence of events seen below.

```
DEBUG 2024-05-02 14:55:06,231 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1197 - [{{0.0.0.0:0}}] accept_batch_start skip because record batch type is batch_type::raft_configuration
DEBUG 2024-05-02 14:55:06,231 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1271 - [{{0.0.0.0:0}}] skip_batch_start called for 1398
DEBUG 2024-05-02 14:55:06,231 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1288 - added offset translation gap [1398-1398], current state: {{base offset/delta: {1397}/1221, map size: 1, last delta: 1221}}
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1197 - [{{0.0.0.0:0}}] accept_batch_start skip because record batch type is batch_type::archival_metadata
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1271 - [{{0.0.0.0:0}}] skip_batch_start called for 1399
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1288 - added offset translation gap [1399-1400], current state: {{base offset/delta: {1397}/1221, map size: 2, last delta: 1222}}
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1197 - [{{0.0.0.0:0}}] accept_batch_start skip because record batch type is batch_type::archival_metadata
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1271 - [{{0.0.0.0:0}}] skip_batch_start called for 1401
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1288 - added offset translation gap [1401-1401], current state: {{base offset/delta: {1397}/1221, map size: 3, last delta: 1224}}
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1197 - [{{0.0.0.0:0}}] accept_batch_start skip because record batch type is batch_type::tx_fence
DEBUG 2024-05-02 14:55:06,232 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1271 - [{{0.0.0.0:0}}] skip_batch_start called for 1402
DEBUG 2024-05-02 14:55:06,234 [shard 2:fetc] cloud_storage - [fiber848~132~149~0 kafka/topic/0] - remote_segment.cc:1185 - [{{0.0.0.0:0}}] accept_batch_start stop parser because 1403 > 177(kafka offset)
DEBUG 2024-05-02 14:55:06,234 [shard 2:fetc] cloud_storage - [fiber41664 kafka/topic/0] - remote_partition.cc:435 - No results, current rp offset: 1403, current kafka offset: 178, max rp offset: 177
DEBUG 2024-05-02 14:55:06,234 [shard 2:fetc] cloud_storage - [fiber848~132~149 kafka/topic/0] - remote_segment.cc:1476 - [{{0.0.0.0:0}}] remote_segment_batch_reader::stop
DEBUG 2024-05-02 14:55:06,234 [shard 2:fetc] cloud_storage - [fiber848~132~149 kafka/topic/0] - remote_segment.cc:1482 - [{{0.0.0.0:0}}] remote_segment_batch_reader::stop - parser-close
DEBUG 2024-05-02 14:55:06,234 [shard 2:main] cloud_storage - [fiber41664 kafka/topic/0] - remote_partition.cc:247 - abort requested via config.abort_source
DEBUG 2024-05-02 14:55:06,234 [shard 2:fetc] cloud_storage - [fiber41665 defa47c0/kafka/topic/0_18762/1398-1405-846-134-v1.log.137] - segment_chunk_data_source.cc:48 - chunk data source destroyed
DEBUG 2024-05-02 14:55:06,235 [shard 2:main] cloud_storage - [fiber848~132~149 kafka/topic/0] - remote_segment.cc:1476 - [{{0.0.0.0:0}}] remote_segment_batch_reader::stop
DEBUG 2024-05-02 14:55:06,235 [shard 2:main] cloud_storage - [fiber848~132~149 kafka/topic/0] - remote_segment.cc:1482 - [{{0.0.0.0:0}}] remote_segment_batch_reader::stop - parser-close
ERROR 2024-05-02 14:55:06,235 [shard 2:fetc] assert - Assert failure: (/var/lib/buildkite-agent/builds/buildkite-arm64-builders-i-009787d74859cb582-1/redpanda/redpanda/src/v/storage/segment_reader.cc:210) '_parent == nullptr' Must close before destroying
```

Note, the tx_fence is assigned a Kafka offset 178, but isn't reflected by the LSO-1=177 calculation used as an upper bound for the log reader config, hence the early return.

This commit fixes the issue by stopping the segment reader before stopping, ensuring only one concurrent caller of stop().

This also adds test simulates a failure where a client disconnect racing with the natural stopping of a segment reader led to a crash. Without the fix, and adding some sleeps in remote_segment_batch_reader::stop(), this would crash fairly consistently, albeit with slightly different backtraces than reported in the issue.

Fixes #18213

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.1.x
- [x] v23.3.x
- [x] v23.2.x

## Release Notes

### Bug Fixes

* Fixes a crash caused by a race between a client disconnect and a segment reader in tiered storage.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/48645#018f3bf2-14de-4f3f-9247-1909bd286845:
```
"rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.ABS"
```
<!--90f87491-9c63-49b5-9ca6-ba67f7bd12b4-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/48645#018f3bf2-14e2-4e58-9794-9fada2637857
<!--cc6f03a0-9e97-47a0-886f-5b62cc397e85-->

dotnwat reviewed (approved):
seems like a pretty good analysis. looking back at john's comments here https://github.com/redpanda-data/redpanda/pull/7800 it also seems like stop() being closed multiple times was a concern in the past, but the fix with `_stopped` didn't seem to fully address it?

would be great to see if @abhijat has any memory of this.

piyushredpanda commented:
Known failure: https://github.com/redpanda-data/redpanda/issues/17535

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
/backport v23.3.x

vbotbuildovich commented:
/backport v23.2.x

vbotbuildovich commented:
Failed to create a backport PR to v23.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-18229-v23.2.x-403 remotes/upstream/v23.2.x
git cherry-pick -x 226c43e52a31ab41188f9ff0ac616e7bf9b42e2d
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/8936989687).

abhijat commented:
> seems like a pretty good analysis. looking back at john's comments here #7800 it also seems like stop() being closed multiple times was a concern in the past, but the fix with `_stopped` didn't seem to fully address it?
> 
> would be great to see if @abhijat has any memory of this.

#7800 was mostly to address issues with remote segments themselves stopping twice, we also added a fix in https://github.com/redpanda-data/redpanda/pull/7868 to address the actual issue, #7800 is mostly defensive. That seems to not have re-occurred here.

IIRC the problem (not relevant to this issue directly but similar in nature) there was that the remote partition while stopping would stop all its chiled segments and `materialized segments` could decide to evict that same segment while trimming at the same time, stopping the same segment twice. So we first removed the remote partition's owned segments using the list hook and moved the segments to stop into a temp variable before stopping.

andrwng commented:
> > seems like a pretty good analysis. looking back at john's comments here #7800 it also seems like stop() being closed multiple times was a concern in the past, but the fix with `_stopped` didn't seem to fully address it?
> > would be great to see if @abhijat has any memory of this.
> 
> #7800 was mostly to address issues with remote segments themselves stopping twice, we also added a fix in #7868 to address the actual issue, #7800 is mostly defensive. That seems to not have re-occurred here.
> 
> IIRC the problem (not relevant to this issue directly but similar in nature) there was that the remote partition while stopping would stop all its chiled segments and `materialized segments` could decide to evict that same segment while trimming at the same time, stopping the same segment twice. So we first removed the remote partition's owned segments using the list hook and moved the segments to stop into a temp variable before stopping.

Thanks for the details! Yeah the fact that background closing is serialized makes the one-stop rule a bit easier to verify.
